File: src/bin/0077-combinations.cpp
// @leetup=info id=77 lang=cpp slug=combinations

#include <algorithm>
#include <iostream>
#include <map>
#include <set>
#include <tuple>
#include <vector>

// @leetup=code

class Solution {
 public:
  std::vector<std::vector<int>> res;
  std::vector<int> s;
  int n;

  std::vector<std::vector<int>> combine(int n, int k) {
    this->n = n;
    dfs(0, k);
    return res;
  }

  void dfs(int pick, int remain) {
    if (remain == 0) {
      res.push_back(s);
      return;
    }
    if (pick == n) {
      return;
    }

    // Pick
    s.push_back(pick + 1);
    dfs(pick + 1, remain - 1);
    s.pop_back();

    // Don't pick
    dfs(pick + 1, remain);
  }
};
// @leetup=code

int main() {
  std::cout << "Hello, world!" << std::endl;
  return 0;
}

File: src/bin/0077-combinations.py
# @leetup=info id=77 lang=python3 slug=combinations



# @leetup=code


class Solution:
    def combine(self, n: int, k: int) -> list[list[int]]:
        res = []
        combi = []

        def dfs(picked: int, remain: int):
            if remain == 0:
                res.append(combi[:])
                return
            if picked == n:
                return

            # pick
            combi.append(picked + 1)
            dfs(picked + 1, remain - 1)
            combi.pop()

            # skip
            dfs(picked + 1, remain)

        dfs(0, k)
        return res


# @leetup=code
